File: maximal-rectangle.java
public class Solution {

    private static class Bar {
        int index;
        int height;

        public Bar(int i, int h) {
            this.index = i;
            this.height = h;
        }
    }

    public int maximalRectangle(char[][] matrix) {
        if (matrix == null || matrix.length == 0) return 0;
        int rows = matrix.length;
        int cols = matrix[0].length;
        // Compute heights.
        int[][] height = new int[rows][cols];
        for (int i = 0; i < cols; i++)
            height[0][i] = matrix[0][i] - '0';
        for (int i = 1; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                height[i][j] = matrix[i][j] == '0' ? 0 : height[i - 1][j] + 1;
            }
        }
        int ans = 0;
        for (int i = 0; i < rows; i++) {
            Stack<Bar> stack = new Stack<Bar>();
            for (int j = 0; j < cols; j++) {
                while (!stack.empty() && stack.peek().height > height[i][j]) {
                    Bar top = stack.pop();
                    int left = stack.empty() ? 0 : stack.peek().index + 1;
                    ans = Math.max(ans, (j - left) * top.height);
                }
                stack.push(new Bar(j, height[i][j]));
            }
            while (!stack.empty()) {
                Bar top = stack.pop();
                int left = stack.empty() ? 0 : stack.peek().index + 1;
                ans = Math.max(ans, (cols - left) * top.height);
            }
        }
        return ans;
    }
}

File: maximal-rectangle.cc
#include <stack>
#include <vector>
using namespace std;

class Solution {
public:
  int maximalRectangle(vector<vector<char> > &matrix) {
    if (matrix.empty() || matrix[0].empty()) return 0;
    int n = matrix.size();
    int m = matrix[0].size();
    // Compute the height of max possible rectangles from a grid.
    vector<vector<int> > height(n, vector<int>(m));
    for (int i = 0; i < m; i++) {
      height[0][i] = matrix[0][i] == '1';
    }
    for (int i = 1; i < n; i++) {
      for (int j = 0; j < m; j++) {
        height[i][j] = matrix[i][j] == '1' ? height[i - 1][j] + 1 : 0;
      }
    }
    int ans = 0;
    for (int i = 0; i < n; i++) {
      stack<Beam> beams;
      for (int j = 0; j < m; j++) {
        while (!beams.empty() && beams.top().height > height[i][j]) {
          Beam last_beam = beams.top();
          beams.pop();
          int left_pos = beams.empty() ? 0 : beams.top().position + 1;
          int area = last_beam.height * (j - left_pos);
          if (area > ans) ans = area;
        }
        beams.push(Beam(j, height[i][j]));
      }
      while (!beams.empty()) {
        Beam last_beam = beams.top();
        beams.pop();
        int left_pos = beams.empty() ? 0 : beams.top().position + 1;
        int area = last_beam.height * (m - left_pos);
        if (area > ans) ans = area;
      }
    }
    return ans;
  }

  struct Beam {
    int position;
    int height;
    Beam(int pos, int h) : position(pos), height(h) {}
  };
};
